### DISCUSSION TITLE
any tips on debugging

### DISCUSSION THREAD

**Discussion by foldl** [rjmacarthy/twinny#248] [General]
I am writing [an OpenAI API compatible server](https://github.com/foldl/chatllm.cpp/blob/master/bindings/openai_api.py) for `twinny`. After enabled debugging, I can only see information about the request in Developer Tool.

The issue I encounter is that streamed FIM response is not recognized by `twinny`. Any tips on how to debug this?

**Comment by rjmacarthy** [comment]
Hello, the best way to debug it is to build it from source.  

1. Clone the repository, 
2. Run `npm i` inside the root dir
3. Open the dir it inside vscode
4. Press `f5` to start debugging
5. Open `stream.ts` set some breakpoints on the response/request

Many thanks.

**Comment by foldl** [comment]
Thanks for your suggestions. I am scared of using `npm`, 😄 Fortunately, chatllm.cpp is OK when setting provider to `litellm`.

From `utils.ts`, I see that there are subtle differences in different providers. Could `twinny` add a generic `OpenAI Compatible` provider? I think `litellm` (and my chatllm.cpp) belongs to this category.
